src/VisualStudio/Core/Def/Implementation/Workspace/SourceGeneratedFileManager.cs

```diff
@@ -0,0 +1,422 @@
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.Imaging.Interop;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Threading;
+using Roslyn.Utilities;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation
+{
+    /// <summary>
+    /// Provides the support for opening files pointing to source generated documents, and keeping the content updated accordingly.
+    /// </summary>
+    [Export(typeof(SourceGeneratedFileManager))]
+    internal sealed class SourceGeneratedFileManager : IRunningDocumentTableEventListener
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly IThreadingContext _threadingContext;
+        private readonly IVsRunningDocumentTable _runningDocumentTable;
+        private readonly ITextDocumentFactoryService _textDocumentFactoryService;
+
+        private readonly RunningDocumentTableEventTracker _runningDocumentTableEventTracker;
+
+        /// <summary>
+        /// The temporary directory that we'll create file names under to act as a prefix we can later recognize and use.
+        /// </summary>
+        private readonly string _temporaryDirectory;
+
+        /// <summary>
+        /// Map of currently open generated files; the key is the generated full file path.
+        /// </summary>
+        private readonly Dictionary<string, OpenSourceGeneratedFile> _openFiles = new Dictionary<string, OpenSourceGeneratedFile>();
+        private readonly VisualStudioWorkspace _visualStudioWorkspace;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public SourceGeneratedFileManager(
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IThreadingContext threadingContext,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            ITextDocumentFactoryService textDocumentFactoryService,
+            VisualStudioWorkspace visualStudioWorkspace)
+        {
+            _serviceProvider = serviceProvider;
+            _threadingContext = threadingContext;
+            _textDocumentFactoryService = textDocumentFactoryService;
+            _temporaryDirectory = Path.Combine(Path.GetTempPath(), "VisualStudioSourceGeneratedDocuments");
+            _visualStudioWorkspace = visualStudioWorkspace;
+
+            Directory.CreateDirectory(_temporaryDirectory);
+
+            _runningDocumentTable = _serviceProvider.GetService<SVsRunningDocumentTable, IVsRunningDocumentTable>();
```

Comment:
❕ No UI thread dependencies in MEF constructors (including COM usage)

Reply:
The shell has a special case here: it is guaranteed to be free threaded enough that you can request it and advise events. I'll add a comment since we've written that comment in other places where we do this same thing.

Reply:
Comment has been added.